### DISCUSSION TITLE
Understanding teleport network flow paths

### DISCUSSION THREAD

**Discussion by infogulch** [gravitational/teleport#13834] [Q&A] [answered]
Hi! I'm new to Teleport and I'm trying to understand how you might expect to configure network access between various components of a classic backend+database app architecture.

I've created a simple [mermaid](https://mermaid-js.github.io/mermaid/#/flowchart) [diagram](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) of what I think it might look like, and have some questions about it below.

```mermaid
graph TD
user{End user} -- HTTPS --> proxy;
admin{Admin user} -- HTTPS/SSH/DB --> teleportproxy

subgraph Internal Network
proxy{App proxy} -- HTTP -->apphttp
teleportproxy{Teleport proxy} -- HTTP --> appdebug
teleportproxy -- SSH --> appssh
teleportproxy -- DB --> dbproto
teleportproxy -- SSH --> dbssh

teleportproxy --> teleportadmin{Teleport Admin}
teleportadmin --> userdirectory{Directory}

  subgraph App Server 1 of N
  apphttp{App Endpoint}
  appdebug{App Debug Port}
  appssh{App SSH Port}
  end

  subgraph DB Server
  dbproto{DB Protocol Port}
  dbssh{DB SSH Port}
  apphttp -- DB --> dbproto
  end
end
```

<details><summary>Mermaid source</summary>

```
graph TD
user{End user} -- HTTPS --> proxy;
admin{Admin user} -- HTTPS/SSH/DB --> teleportproxy

subgraph Internal Network
proxy{App proxy} -- HTTP -->apphttp
teleportproxy{Teleport proxy} -- HTTP --> appdebug
teleportproxy -- SSH --> appssh
teleportproxy -- DB --> dbproto
teleportproxy -- SSH --> dbssh

teleportproxy --> teleportadmin{Teleport Admin}
teleportadmin --> userdirectory{Directory}

  subgraph App Server 1 of N
  apphttp{App Endpoint}
  appdebug{App Debug Port}
  appssh{App SSH Port}
  end

  subgraph DB Server
  dbproto{DB Protocol Port}
  dbssh{DB SSH Port}
  apphttp -- DB --> dbproto
  end
end
```

</details>

1. Would you expect the `App proxy` to be separate (as pictured) from the `Teleport proxy`, or should all the traffic go through the teleport proxy? What designs are recommended and what are supported?
2. Where do Teleport Nodes come in? Would it be a service installed on the `App Server` and `DB Server`?
3. Would Admin traffic flow from the `Teleport proxy` to the `App Server`'s and `DB Server`'s ports directly, or would traffic be mediated by the Teleport Node service running on the servers (see 2)?
4. Can the traffic between the `App Endpoint` and `DB Protocol Port` be mediated or authenticated with Teleport? Should it?
5. Which of these connections if any are authenticated with the teleport-managed CA?

**Comment by webvictim** [comment, marked as answer]
_1. Would you expect the App proxy to be separate (as pictured) from the Teleport proxy, or should all the traffic go through the teleport proxy? What designs are recommended and what are supported?_

I think that ultimately, the answer depends on whether you're looking to secure all app/database access via Teleport, or just individual users accessing those resources using their SSO identity.

Teleport is made up of services like `proxy_service`, `app_service`, `ssh_service` etc which all enable different protocols/functionality within Teleport. You can run as many or as few of these services in one `teleport` process as you like.

You can either deploy a central proxy/app/database service which will then make outbound connections to your app/database hosts (as in your diagram), or you can deploy the `app_service` on the app host itself and have it reach back out to Teleport using a reverse tunnel. Same for databases.


_2. Where do Teleport Nodes come in? Would it be a service installed on the App Server and DB Server?_

SSH is a little different in that if you run an `ssh_service` it's essentially an `sshd` replacement allowing you to log into that one host. You can also run an `app_service` or `db_service` inside the same process using a reverse tunnel to connect back if you want.

You can also configure `sshd` to trust Teleport-issued certs and use this for connections too, without having to run a Teleport `ssh_service` at all. That's the way it works in your diagram too.


_3. Would Admin traffic flow from the Teleport proxy to the App Server's and DB Server's ports directly, or would traffic be mediated by the Teleport Node service running on the servers (see 2)?_

The traffic would flow direct in either case - the node/SSH service has no bearing on the traffic flow for other protocols.


_4. Can the traffic between the App Endpoint and DB Protocol Port be mediated or authenticated with Teleport? Should it?_

Assuming that your application can use TLS certificates for DB authentication, you could theoretically use [Machine ID](https://goteleport.com/docs/machine-id/introduction/) to secure the app -> database connection with short-lived certificates. We don't have a guide written on how to do that just yet, though.


_5. Which of these connections if any are authenticated with the teleport-managed CA?_

- Any Teleport -> Teleport traffic is secured using mutual TLS.
- Teleport db_service -> database traffic is also secured in this way (using a Teleport-issued certificate installed at the database server level).
- Teleport -> sshd traffic can also be secured using a Teleport-issued SSH host certificate.
- Teleport app_service -> application traffic doesn't use Teleport-issued certs - it builds an HTTPS trust chain using the CA certificates pre-installed on the machine, similar to `curl` or a web browser.

**Reply by infogulch** [reply]
Thank you for your detailed answer!
